portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -186,4 +210,23 @@ private String getUploadedFileUrl(HttpServletRequest request, String formInputEl
       }
     }
 
+    // This mimics the doPost method above and is only for testing purposes.
+    public void doPost_ForTests(String msg, String nickname, String tag) {
```

Comment:
This shouldn't be here. If you are going to make utility functions, they should be in their own file. Further, this should be called by doPost and it should be tested before it is submitted.